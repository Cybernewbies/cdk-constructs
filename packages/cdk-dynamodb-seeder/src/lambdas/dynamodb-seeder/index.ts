import {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceCreateEvent,
} from 'aws-lambda';
import { DynamoDB, S3 } from 'aws-sdk';
import chunk from 'lodash.chunk';

// DynamoDB has a 25 item limit in batch requests
// https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html
const MAX_BATCH_CHUNK = 25;

interface DynamoDBSeederProps {
  tableName: string;
  seeds: {
    inlineSeeds?: string;
    s3Bucket?: string;
    s3Key?: string;
    s3ObjectVersion?: string;
  };
}

type Seeds = Record<string, unknown>[];

const dynamodb = new DynamoDB.DocumentClient();
const s3 = new S3();

const getProperties = (
  props: CloudFormationCustomResourceEvent['ResourceProperties'],
): DynamoDBSeederProps => ({
  tableName: props.TableName,
  seeds: {
    inlineSeeds: props.Seeds.InlineSeeds,
    s3Bucket: props.Seeds.S3Bucket,
    s3Key: props.Seeds.S3Key,
    s3ObjectVersion: props.Seeds.S3ObjectVersion,
  },
});

const onCreate = async (
  event: CloudFormationCustomResourceCreateEvent,
): Promise<void> => {
  const props = getProperties(event.ResourceProperties);

  const { inlineSeeds, ...s3Location } = props.seeds;

  const seeds = inlineSeeds
    ? JSON.parse(inlineSeeds)
    : await getSeedsFromS3(s3Location);

  await writeSeeds(props.tableName, seeds);

  console.log(`Seed running complete for table ${props.tableName}`);
};

const getSeedsFromS3 = async (s3Location: {
  s3Bucket?: string;
  s3Key?: string;
  s3ObjectVersion?: string;
}): Promise<Seeds> => {
  const { s3Bucket, s3Key, s3ObjectVersion } = s3Location;

  if (!s3Bucket || !s3Key) {
    throw new Error('Bucket configuration missing!');
  }

  const { Body: body } = await s3
    .getObject({
      Bucket: s3Bucket,
      Key: s3Key,
      VersionId: s3ObjectVersion,
    })
    .promise();

  return JSON.parse(body.toString()) as Seeds;
};

const writeSeeds = async (tableName: string, seeds: Seeds): Promise<void> => {
  const seedChunks = chunk(seeds, MAX_BATCH_CHUNK);

  console.log(`Sending data to dynamodb: ${seedChunks.length} chunks`);

  await Promise.all(
    seedChunks.map(async (seedChunk) => {
      const requests = seedChunk.map((seed) => ({
        PutRequest: {
          Item: seed,
        },
      }));

      return dynamodb
        .batchWrite({
          RequestItems: {
            [tableName]: requests,
          },
        })
        .promise();
    }),
  );
};

export const handler = async (
  event: CloudFormationCustomResourceEvent,
): Promise<void> => {
  const requestType = event.RequestType;

  switch (requestType) {
    case 'Create':
      return onCreate(event as CloudFormationCustomResourceCreateEvent);
    case 'Update':
      return;
    case 'Delete':
      return;
    default:
      throw new Error(`Invalid request type: ${requestType}`);
  }
};
