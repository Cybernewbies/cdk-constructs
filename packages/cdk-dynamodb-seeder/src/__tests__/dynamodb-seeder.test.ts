import * as path from 'path';
import { Stack } from '@aws-cdk/core';
import { AttributeType, Table } from '@aws-cdk/aws-dynamodb';
import 'jest-cdk-snapshot';

import { DynamoDBSeeder } from '../dynamodb-seeder';
import { Seeds } from '../seeds';

jest.mock('../directories', () => ({
  dynamodbSeederDir: path.join(__dirname, 'mocks', 'dynamodb-seeder'),
}));

test('default setup', () => {
  // GIVEN
  const stack = new Stack();
  const table = new Table(stack, 'Table', {
    partitionKey: {
      name: 'id',
      type: AttributeType.NUMBER,
    },
  });

  // WHEN
  new DynamoDBSeeder(stack, 'DynamoDBSeeder', {
    table,
    seeds: Seeds.fromInline([
      {
        id: 1,
        column: 'foo',
      },
      {
        id: 2,
        column: 'bar',
      },
    ]),
  });

  // THEN
  expect(stack).toMatchCdkSnapshot();
});
