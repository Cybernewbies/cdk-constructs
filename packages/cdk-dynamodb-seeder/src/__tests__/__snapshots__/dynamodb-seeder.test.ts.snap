// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default setup 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0ArtifactHashD7FD838A": Object {
      "Description": "Artifact hash for asset \\"4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0\\"",
      "Type": "String",
    },
    "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0S3Bucket7A9635D5": Object {
      "Description": "S3 bucket for asset \\"4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0\\"",
      "Type": "String",
    },
    "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0S3VersionKeyD76530F8": Object {
      "Description": "S3 key for asset version \\"4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomDynamodbSeederCustomResourceProviderHandler904BFFD4": Object {
      "DependsOn": Array [
        "CustomDynamodbSeederCustomResourceProviderRoleB3FCC5E2",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0S3Bucket7A9635D5",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0S3VersionKeyD76530F8",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4c20502c8ee54d52fa9b2e8f0251e77c489e44933dcde320ab40ffdf3a0ad8b0S3VersionKeyD76530F8",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomDynamodbSeederCustomResourceProviderRoleB3FCC5E2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomDynamodbSeederCustomResourceProviderRoleB3FCC5E2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "dynamodb:BatchWriteItem",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::GetAtt": Array [
                      "TableCD117FA1",
                      "Arn",
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DynamoDBSeederCustomResourceEF4D5C60": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Seeds": Object {
          "InlineSeeds": "[{\\"id\\":1,\\"column\\":\\"foo\\"},{\\"id\\":2,\\"column\\":\\"bar\\"}]",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomDynamodbSeederCustomResourceProviderHandler904BFFD4",
            "Arn",
          ],
        },
        "TableName": Object {
          "Ref": "TableCD117FA1",
        },
      },
      "Type": "Custom::DynamodbSeeder",
      "UpdateReplacePolicy": "Delete",
    },
    "TableCD117FA1": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "id",
            "AttributeType": "N",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
